osquery/examples/example_writable_table.cpp

```diff
@@ -0,0 +1,299 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <osquery/core/conversions.h>
+#include <osquery/sdk.h>
+#include <osquery/system.h>
+
+#include <rapidjson/document.h>
+
+#include <mutex>
+#include <sstream>
+
+using namespace osquery;
+class WritableTable : public TablePlugin {
+ private:
+  /// Simple primary key implementation; this is just the two columns
+  /// concatenated
+  using PrimaryKey = std::string;
+
+  /// A rowid value uniquely identifies a row in a table
+  using RowID = std::string;
+
+  /// Data mutex
+  std::mutex mutex;
+
+  /// This is our data; each row contains a rowid, and the two remaining columns
+  /// ('text' and 'integer')
+  std::unordered_map<PrimaryKey, Row> data;
+
+  /// This is used to map rowids to primary keys
+  std::unordered_map<RowID, PrimaryKey> rowid_to_primary_key;
+
+  /// This is an example implementation for a basic primary key
+  PrimaryKey getPrimaryKey(const Row& row) const {
+    return row.at("text") + row.at("integer");
+  }
+
+  /// Returns true if the given primary key is unique; used to adhere to
+  /// constraints
+  bool isPrimaryKeyUnique(
+      const PrimaryKey& primary_key,
+      const std::string& ignored_rowid = std::string()) const {
+    auto it = data.find(primary_key);
+    if (it == data.end()) {
+      return true;
+    }
+
+    if (ignored_rowid.empty()) {
+      return false;
+    }
+
+    return it->second.at("rowid") == ignored_rowid;
+  }
+
+  /// Generates a new rowid value; used when sqlite3 does not provide one
+  size_t generateRowId() const {
+    static size_t rowid_generator = 0U;
+    return rowid_generator++;
+  }
+
+  /// Saves the given row
+  Status saveRow(const Row& row, PrimaryKey primary_key = std::string()) {
+    // Expect full rows (i.e. must include the rowid column)
+    if (row.size() != 3U) {
+      return Status(1, "Invalid column count");
+    }
+
+    // Compute the primary key if we haven't received one
+    if (primary_key.empty()) {
+      primary_key = getPrimaryKey(row);
+    }
+
+    // Save the row and update the index
+    data.insert({primary_key, row});
+
+    const auto& rowid = row.at("rowid");
+    rowid_to_primary_key.insert({rowid, primary_key});
+
+    return Status(0, "OK");
+  }
+
+  /// Expands a value list returned by osquery into a Row (without the rowid
+  /// column)
+  Status getRowData(Row& row, const std::string& value_list) const {
```

Comment:
I kept value_list because it is more descriptive than just 'json' (it is in fact a list of values, regardless of the format). Maybe we can use json_value_array